File: app/src/main/java/in/woloo/www/login/receiver/SmsBroadcastReceiver.java
package in.woloo.www.login.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import com.google.android.gms.auth.api.phone.SmsRetriever;
import com.google.android.gms.common.api.CommonStatusCodes;
import com.google.android.gms.common.api.Status;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import in.woloo.www.utils.Logger;

public class SmsBroadcastReceiver extends BroadcastReceiver {

    private static final String TAG = "SmsBroadcastReceiver";
    OtpReceivedInterface otpReceiveInterface = null;
    public void setOnOtpListeners(OtpReceivedInterface otpReceiveInterface) {
        this.otpReceiveInterface = otpReceiveInterface;
    }
    @Override public void onReceive(Context context, Intent intent) {
        Logger.d(TAG, "onReceive: ");
        if (SmsRetriever.SMS_RETRIEVED_ACTION.equals(intent.getAction())) {
            Bundle extras = intent.getExtras();
            Status mStatus = (Status) extras.get(SmsRetriever.EXTRA_STATUS);
            switch (mStatus.getStatusCode()) {
                case CommonStatusCodes.SUCCESS:
                    // Get SMS message contents'
                    String message = (String) extras.get(SmsRetriever.EXTRA_SMS_MESSAGE);
                    Logger.d(TAG, "onReceive: failure "+message);
                    if (otpReceiveInterface != null) {
                        Pattern pattern = Pattern.compile("(\\d{4})");
                        Matcher matcher = pattern.matcher(message.toString());
                        String otp1 = "";
                        if (matcher.find()) {
                            otp1 = matcher.group(0);  // 4 digit number
                        }
                        otpReceiveInterface.onOtpReceived(otp1);
                    }
                    break;
                case CommonStatusCodes.TIMEOUT:
                    // Waiting for SMS timed out (5 minutes)
                    Logger.d(TAG, "onReceive: failure");
                    if (otpReceiveInterface != null) {
                        otpReceiveInterface.onOtpTimeout();
                    }
                    break;
            }
        }
    }



}

File: app/src/main/java/in/woloo/www/application_kotlin/receiver/SmsBroadcastReceiver.kt
package `in`.woloo.www.application_kotlin.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import com.google.android.gms.auth.api.phone.SmsRetriever
import com.google.android.gms.common.api.CommonStatusCodes
import com.google.android.gms.common.api.Status
import `in`.woloo.www.application_kotlin.interfaces.OtpReceivedInterface
import `in`.woloo.www.utils.Logger
import java.util.regex.Pattern

class SmsBroadcastReceiver : BroadcastReceiver() {
    private var otpReceiveInterface: OtpReceivedInterface? = null
    fun setOnOtpListeners(otpReceiveInterface: OtpReceivedInterface?) {
        this.otpReceiveInterface = otpReceiveInterface
    }

    override fun onReceive(context: Context, intent: Intent) {
        Logger.d(TAG, "onReceive: ")
        if (SmsRetriever.SMS_RETRIEVED_ACTION == intent.action) {
            val extras = intent.extras
            val mStatus = extras!![SmsRetriever.EXTRA_STATUS] as Status?
            when (mStatus!!.statusCode) {
                CommonStatusCodes.SUCCESS -> {
                    // Get SMS message contents'
                    val message = extras[SmsRetriever.EXTRA_SMS_MESSAGE] as String?
                    Logger.d(TAG, "onReceive: failure $message")
                    if (otpReceiveInterface != null) {
                        val pattern = Pattern.compile("(\\d{4})")
                        val matcher = pattern.matcher(message.toString())
                        var otp1: String? = ""
                        if (matcher.find()) {
                            otp1 = matcher.group(0) // 4 digit number
                        }
                        otpReceiveInterface!!.onOtpReceived(otp1)
                    }
                }

                CommonStatusCodes.TIMEOUT -> {
                    // Waiting for SMS timed out (5 minutes)
                    Logger.d(TAG, "onReceive: failure")
                    if (otpReceiveInterface != null) {
                        otpReceiveInterface!!.onOtpTimeout()
                    }
                }
            }
        }
    }

    companion object {
        private const val TAG = "SmsBroadcastReceiver"
    }
}
